x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { useState, useEffect, useRef, useCallback } from 'react';
+import { isEqual } from 'lodash';
+import {
+  Map,
+  MapboxOptions,
+  NavigationControl,
+  Popup,
+  GeoJSONSourceOptions
+} from 'mapbox-gl';
+import { GeoJsonProperties } from 'geojson';
+import euiLightVars from '@elastic/eui/dist/eui_theme_light.json';
+import { shade, tint } from 'polished';
+
+interface ChoroplethDataElement {
+  key: string;
+  value: number;
+  [property: string]: any;
+}
+
+interface Props {
+  style?: React.CSSProperties;
+  initialGeojsonSource: NonNullable<GeoJSONSourceOptions['data']>;
+  geojsonKeyProperty: string;
+  data: ChoroplethDataElement[];
+  renderTooltip: (props: {
+    geojsonProperties: NonNullable<mapboxgl.MapboxGeoJSONFeature['properties']>;
+    data?: ChoroplethDataElement;
+  }) => React.ReactElement | null;
+  initialMapboxOptions?: Partial<MapboxOptions>;
+}
+
+const CHOROPLETH_LAYER_ID = 'choropleth_layer';
+const CHOROPLETH_POLYGONS_SOURCE_ID = 'choropleth_polygons';
+
+export function getProgressionColor(scale: number) {
+  const baseColor = euiLightVars.euiColorPrimary;
+  const adjustedScale = 0.75 * scale + 0.05; // prevents pure black & white as min/max colors.
+  if (adjustedScale < 0.5) {
+    return tint(adjustedScale * 2, baseColor);
+  }
+  if (adjustedScale > 0.5) {
+    return shade(1 - (adjustedScale - 0.5) * 2, baseColor);
+  }
+  return baseColor;
+}
+
+export function getDataRange(data: Props['data']) {
+  const firstValue = data[0] ? data[0].value : 0;
+  return data.reduce(
+    ([min, max], { value }) => [Math.min(min, value), Math.max(max, value)],
+    [firstValue, firstValue]
+  );
+}
+
+export const ChoroplethMap: React.SFC<Props> = props => {
+  const {
+    style,
+    initialGeojsonSource,
+    geojsonKeyProperty,
+    data,
+    renderTooltip: ToolTip,
+    initialMapboxOptions
+  } = props;
+
+  const containerRef = useRef<HTMLDivElement>(null);
+  const [map, setMap] = useState<Map | null>(null);
+  const popupRef = useRef<Popup | null>(null);
+  const popupContainerRef = useRef<HTMLDivElement>(null);
+  const enableScrollZoom = useRef(false);
+  const [hoverState, setHoverState] = useState<{
+    geojsonProperties?: GeoJsonProperties;
+    data?: ChoroplethDataElement;
+  }>({});
```

Comment:
Isn't this only used for the tooltip? If so, then what about calling it `tooltipState` and only set the things you need:
```js
  const [tooltipState, setTooltipState] = useState<{
  name: string
  value: number; 
  docCount: number;
  } | null>(null);
```

I think this will simplify a lot of other things.

Reply:
Right now we are the only consume so I think we should make things as simple as possible for ourselves.